### DISCUSSION TITLE
[CodeStyle] Namespaces

### DISCUSSION THREAD

**Discussion by ViMaSter** [Studio-Lovelies/GG-JointJustice-Unity#250] [Ideas]
While typing out my response for https://github.com/Studio-Lovelies/GG-JointJustice-Unity/pull/245#discussion_r841127339, I noticed, that namespaces are currently used inconsistently in our codebase (or rather: I started using them out of habit and never realized that we're not actually using them; oops.)

Unfortunately, it doesn't seem like [Unity offers any recommendations or best practices.](https://docs.unity3d.com/Manual/Namespaces.html)

I see a few benefits in using them:
  - Immediate insight into which code belongs to which assembly (if we have completed https://github.com/Studio-Lovelies/GG-JointJustice-Unity/issues/173 and code split into "things we need in the final game" [`Game.`], "things we use as developers for testing" [`Demo.`], "thing automated tests use") [`*ModeTests.`]
  - Personally, I mostly navigate our codebase [by just fuzzy-searching](https://i.vimaster.de/zMiklWUQqQ.png) a class I want to work on. If you'd ask me to write out our current folder structure, I probably couldn't. Having our folder structure mirror our namespaces would probably make architecture decisions a bit easier, because of the visibility of which code is grouped where.
    - `Game.Story.Playlist` -> `Assets/Scripts/Story/Playlist.cs`
    - `Tests.PlayModeTests.Scripts.AudioControllerTests` -> `Assets/Tests/PlayModeTests/Scripts/AudioControllerTests.cs`

If I would have to put this into actionable items, this is what I could see working:
- As part of https://github.com/Studio-Lovelies/GG-JointJustice-Unity/issues/173, use the following prefixes for code currently in these folders:
  - `Assets/Scripts` -> `Game`
  - `Assets/Demo/Scripts` -> `Demo`
  - `Assets/Tests/[1]/[2]` -> `Tests.[1].[2]`
- Extend our [Code Style Guideline](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/wiki/Code-Style-Guideline) to encourage namespaces when creating multiple classes that heavily rely on themselves
  - I think even discouraging (or disallowing) adding classes to the `global::` namespace (/ not adding a namespace at all) might be a good idea too?
- Extend our [Code Style Guideline](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/wiki/Code-Style-Guideline) to use namespace and class names to hint at the implementation of a design pattern (when applicable)

As usual, curious to hear everyone else's thoughts about it!

**Comment by PhoebeMitchell** [comment]
Generally in my own projects I do try and use namespaces. However I'm really bad at dividing my code into meaningful namespaces and end up with a whole ton of `using`s at the top of my scripts due to scripts in different namespaces communicating with each other.

On the topic of using namespaces in _our_ game, I decided to not bother because almost no Unity guide uses them, and felt it would complicate things as nobody else was using them, and might not know how to use them.

I feel retroactively going back and adding namespaces is a little pointless as it wouldn't be enforcing any kind of separation of code, we'd still have all the classes communicating with each other, just with more code at the top of the files. Moving forward though I feel it could be good.

**Reply by ViMaSter** [reply]
> Generally in my own projects I do try and use namespaces. However, I'm really bad at dividing my code into meaningful namespaces and end up with a whole ton of usings at the top of my scripts due to scripts in different namespaces communicating with each other.

Well, I think practice makes perfect! : D  
Especially with the PR review process, people like Explosive (that actually built entire engines and games in them) might have some valuable insight. Because admittedly, I also only ever ended up using them if I had a project that actually spanned multiple libraries, just so I'm avoiding any `global::` namespace collisions.

> I feel retroactively going back and adding namespaces is a little pointless as it wouldn't be enforcing any kind of separation of code, we'd still have all the classes communicating with each other, just with more code at the top of the files

Just to clarify here: Would your point still apply, even if we only _namespaceify_ our current structure of assemblies (`Game.`/`PlayModeTest.Scripts.`/`PlayModeTest.Scenes.`/`EditModeTest.`)? Or just once we also add the folders? Or are you talking about a scenario, where we try to come up with and retrofit some other structure? (Which - as you said - we don't really have, so why try to come up with artificial guidelines and reasons for it now? 👀)

> Moving forward though I feel it could be good.

I think this is the most important thing in any case, and I heavily agree with it too: There's **no urgent need** for it **just yet**, but I can see use benefitting from it (or being pained by not doing it) in the long run.  
I'm imagining things like minigames or other any other "non-linear narrative gameplay elements" that could live in namespaces and based on that maybe will help us come up with useful divisions of sections of code. (Just to name a few potential candidates: `Game.Narrative.`, `Game.MainMenu.`, `Game.Fishing.`<sup>? I mean we just **_gotta_** have a fishing minigame</sup>, `Game.Settings.`)

**Reply by PhoebeMitchell** [reply]
> Just to clarify here: Would your point still apply, even if we only namespaceify our current structure of assemblies (Game./PlayModeTest.Scripts./PlayModeTest.Scenes./EditModeTest.)? Or just once we also add the folders? Or are you talking about a scenario, where we try to come up with and retrofit some other structure? (Which - as you said - we don't really have, so why try to come up with artificial guidelines and reasons for it now? 👀)

I mean having a namespace for each folder like Evidence, NarrativeScript, ActionDecoder, etc. I feel adding namespaces for these wouldn't be that beneficial at this point, at least not without any refactoring so that our classes are less interwoven.

> Just to clarify here: Would your point still apply, even if we only namespaceify our current structure of assemblies

More general namespacing would be better, however I don't see much purpose in having a namespace for each assembly as the assemblies already keep code in each assembly separate.
